feat: async common revalidate view and api

nikomakela opened the pull request:
LIIKUNTA-702.

Add `NextPageRevalidateApi` class to provide a common interface to revalidate NextJS pages (in NextJS pages-router) on demand. The new `NextPageRevalidateAPI` uses a new `NextPageRevalidateService` that can be configured app wise.

The old implementation was a synchronous handler that waited for response but was always disconnected with a connection timeout. The new handler works asynchronously and should response immediately, but keeps working on background. This commit refactors how the static page paths (for pages and posts/articles) is fetched. Earlier a huge query was made just to get an URI from CMS content, so now the query only fetches the data that it really needs, which makes the quering a lot faster.

The new way to revalidate the generated pages uses filesystem only: it finds the html and json files in NextJS's server folder (for ISR pages), reads their slugs and creates an array of pathnames, that can be used to trigger revalidation for each generated page. If the CMS was used as a source for static paths, it could not be used for venues, courses and events, plus it gave lots of paths to pages that were not generated yet, or were added wrongly to CMS (not under the pages-path) so they could never be generated.

nikomakela committed:
feat: async common revalidate view and api

LIIKUNTA-702.

Add `NextPageRevalidateApi` class to provide a common interface to
revalidate NextJS pages (in NextJS pages-router) on demand. The new
`NextPageRevalidateAPI` uses a new `NextPageRevalidateService` that can
be configured app wise.

The old implementation was a synchronous handler that waited for
response but was always disconnected with a connection timeout. The new
handler works asynchronously and should response immediately, but keeps
working on background.

This commit also refactors how the static page paths (for pages and
posts/articles) is fetched. Earlier a huge query was made just to get an
URI from CMS content, so now the query only fetches the data that it
really needs, which makes the quering a lot faster.

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr781.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela committed:
chore: fix CMS_ORIGIN in env examples

LIIKUNTA-702

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr781.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr781.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr781.dev.hel.ninja :laughing::tada::tada::tada:

nikomakela committed:
feat: revalidate API for courses, events and venues

LIIKUNTA-702.

While using a NextJS with pages-router, there is no simple way to
revalidate all ISR pages, since there is no proper way to get a list of
generated pages. When revalidating the CMS pages (and posts), the URIs
can be fetched from the CMS server and then the list of URIs can be used
to regenerate all the static pages generated for those URIs, but for
events, courses and venues we don't have a source or API to get list of
URIs. For those, there is now another helper class, that tries to check
what html and json files there are in certain selected paths (under
.next/server/pages). Then those filenames and their paths can be used to
create a list of URIs, that can be used to revalidate them.

Add `skipBuildStaticGeneration` as a getter to `AppConfig` to fix the
reading of related environment variable. The env-variable could not be
set to any falsy value, so the process was always skipped build time,
when an environment variable was defined (to any value).

The revalidation API now uses filesystem only for generated page paths.

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr781.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr781.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr781.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr781.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr781.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr781.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr781.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr781.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr781.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr781.dev.hel.ninja :laughing::tada::tada::tada:
